### ISSUE TITLE
Suggestions for ignoring change notification depending on source?

### ISSUE THREAD

**Issue by swiftlyfalling** [groue/GRDB.swift#43]
Thank you for a wonderful library. 👍

My question regarding change notifications is perhaps most easily illustrated by an example.

Example:
An app syncs a local database to iCloud using CloudKit.
The syncing engine registers a TransactionObserver to monitor local database changes, and records them for uploading to CloudKit.
When changes in CloudKit (i.e. from another device) need to be synced back to the local device, what's the best method to ignore them in the syncing engine's TransactionObserver?

i.e. The syncing engine only needs to record (and sync) local database changes made by other parts of the app. So it would want to be able to ignore change notifications caused by its own database transactions.

Is there any suggested way to accomplish this currently? Perhaps some way to tag database write transactions with an app-defined tag/source (that could then be retrieved in the TransactionObserver events)?

**Comment by groue** [comment]
Hello @swiftlyfalling 

What about removing the observer before starting the syncing back from CloudKit, and adding it back after the syncing is done?

**Comment by groue** [comment]
BTW, I'm happy to see TransactionObserver put to some good use :-)

**Comment by groue** [comment]
Hmmm. I was suggesting something like:

```swift
// Bad code, actually
func applyCloudKitChanges(changes) {
    dbPool.removeObserver(observer) // or dbQueue, it's the same
    dbPool.writeInTransaction { db in
        // Apply CloudKit changes
        return .Commit
    }
    dbPool.addObserver(observer)
}
```

The trouble with the code above is that a thread could modify the database after the cloudkit transaction has ended, and before the observer is added back.

So this is not very robust.

A robust solution is to make sure that the observer is configured in the serialized writer queue:

```swift
class MyObserver: TransactionObserverType {
    var needsIgnoreNextTransaction: Bool = false
    
    func databaseDidCommit(db: Database) {
        // We're in the serialized writer queue, here.
        if needsIgnoreNextTransaction {
            needsIgnoreNextTransaction = false
            return
        }
        // Process recorded changes
    }
}

func applyCloudKitChanges(changes) {
    dbPool.writeInTransaction { db in
        // We're in the serialized writer queue, here.
        observer.needsIgnoreNextTransaction = true
        // Apply CloudKit changes
        return .Commit
    }
}
```

Tell me if this fits your situation.

**Comment by swiftlyfalling** [comment]
Had considered the remove / re-add observer route, but was concerned about the concurrency issues you mentioned.

With the clarification that all TransactionObserver callbacks are serialized with all database updates, your second proposal should work well! Will test and get back to you if any concerns pop up.

And again, many thanks for a fantastic framework. 👍

**Comment by groue** [comment]
Thanks @swiftlyfalling, I'm glad GRDB helps you building your application!

**Comment by groue** [comment]
@swiftlyfalling: were you able to successfully tell apart your transaction sources, eventually?

**Comment by swiftlyfalling** [comment]
@groue Haven't had a chance to implement this functionality yet - got stuck on other priorities. Apologies. Will let you know once I have a chance to look into this again.

**Comment by groue** [comment]
No problem. If you have any trouble or question, you'll open another issue. Happy GRDB!
